pkg/terraform/state/handler.go

```diff
@@ -76,136 +73,178 @@ func (h *handler) OnChange(key string, obj *v1.State) (*v1.State, error) {
-			return obj, err
-		}
+	if v1.StateConditionJobDeployed.IsTrue(obj) && obj.Status.LastRunHash != "" {
+		return obj, fmt.Errorf("job already running %s", obj.Status.LastRunHash)
+	}
 
-		if obj.Status.ExecutionName != runName {
-			obj.Status.ExecutionName = runName
-			obj, err = h.states.Update(obj)
-			if err != nil {
-				return obj, errors.WithMessage(err, "updating execution on state change")
-			}
-		}
-		v1.ExecutionConditionDestroyJobDeployed.True(obj)
-		return h.states.Update(obj)
-	})
+	runHash := createRunHash(obj, input, ActionDestroy)
+	if runHash == obj.Status.LastRunHash {
+		logrus.Debug("hashes the same, nothing to do")
+		return obj, fmt.Errorf("job already running %s", obj.Status.LastRunHash)
+	}
 
+	//running an execution
+	logrus.Debug("acquire lock")
+	obj.Status.LastRunHash = runHash
+	obj, err = h.states.Update(obj)
 	if err != nil {
-		return obj, errors.WithMessage(err, "track error")
+		logrus.Debug("lock failed")
+		logrus.Debug(err)
+		return obj, err
 	}
 
-	if runName == "" {
-		combinedVars := combineVars(input)
-		combinedVars["key"] = obj.Name
-		name := createExecRunAndSecretName(obj, combinedVars, input.Module.Status.ContentHash)
-		runName = name
-		logrus.Debugf("Creating destroy job: %s", runName)
+	logrus.Debug("deploy destroy")
+
+	//no job running, try and run a destroy
+	exec, err := h.deployDestroy(obj, input)
+	if err != nil {
+		logrus.Debug(err)
+		return obj, err
 	}
 
-	execution, err := h.executions.Get(obj.Namespace, runName, metaV1.GetOptions{})
+	obj.Status.ExecutionName = exec.Name
+	v1.StateConditionJobDeployed.True(obj)
+
+	//destroy deployed, setup watcher.
+	execution, err := h.executions.Get(obj.Namespace, obj.Status.ExecutionName, metaV1.GetOptions{})
 	if err != nil {
+		logrus.Debug("error getting the execution")
 		return obj, errors.WithMessage(err, "error getting execution")
 	}
 
-	if v1.ExecutionRunConditionApplied.IsTrue(execution) {
-		err = h.states.Delete(execution.Namespace, execution.Name, &metaV1.DeleteOptions{})
-		if err != nil {
-			if !k8serrors.IsNotFound(err) {
-				return obj, err
-			}
-		}
-		return obj, nil
-	}
+	go h.watchExecution(obj, execution, ActionDestroy)
 
-	if !v1.ExecutionConditionWatchRunning.IsTrue(obj) {
-		go h.watchDestroyRun(obj, execution)
-		v1.ExecutionConditionWatchRunning.True(obj)
+	//return error because no error will clear finalizer even if job did not complete.
+	_, err = h.states.Update(obj)
+	if err != nil {
+		return obj, err
 	}
 
-	return h.states.Update(obj)
+	return obj, fmt.Errorf("watching execution to complete")
 }
 
-// watchDestroyRun checks the execution for the Applied condition, once set
-// terraform destroy was run so the state can be requeued for deletion.
-func (h *handler) watchDestroyRun(state *v1.State, execution *v1.Execution) {
+func (h *handler) watchExecution(state *v1.State, execution *v1.Execution, action string) {
+	logrus.Debugf("About to watch %s", execution.Name)
+	maxIterations := 450 //15min
```

Comment:
just write a normal for loop, no reason to create all the same vars outside the loop.

Reply:
agree, but it's going away, see comment above